flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlToOperationConverterTest.java

```diff
@@ -1466,6 +1477,404 @@ public void testAlterTableCompactOnManagedPartitionedTable() throws Exception {
+                operation,
+                tableIdentifier,
+                Schema.newBuilder()
+                        .column("c", DataTypes.BIGINT().notNull())
+                        .column("b", DataTypes.BIGINT().notNull())
+                        .column("a", DataTypes.STRING())
+                        .withComment("move a after b, change nullability, and drop pk")
+                        .primaryKey("c")
+                        .build());
+
+        // test modify physical column to computed column
+        operation =
+                parse(
+                        "alter table tb1 modify c as b + 1 comment 'change c to b + 1'",
+                        SqlDialect.DEFAULT);
+
+        assertAlterTableSchema(
+                operation,
+                tableIdentifier,
+                Schema.newBuilder()
+                        .column("a", DataTypes.STRING().notNull())
+                        .column("b", DataTypes.BIGINT().notNull())
+                        .columnByExpression("c", "`b` + 1")
+                        .withComment("change c to b + 1")
+                        .primaryKeyNamed("ct1", "a", "b")
+                        .build());
+    }
+
+    @Test
+    public void testAlterTableSchemaModifyConstraint() throws Exception {
+        prepareNonManagedTable(true);
+
+        // test modify constraint name
+        Operation operation =
+                parse(
+                        "alter table tb1 modify constraint ct2 primary key (a, b) not enforced",
+                        SqlDialect.DEFAULT);
+        assertAlterTableSchema(
+                operation,
+                ObjectIdentifier.of("cat1", "db1", "tb1"),
+                Schema.newBuilder()
+                        .column("a", DataTypes.STRING().notNull())
+                        .column("b", DataTypes.BIGINT().notNull())
+                        .column("c", DataTypes.BIGINT())
+                        .primaryKeyNamed("ct2", "a", "b")
+                        .build());
+
+        // test modify pk will change column nullability
+        operation =
+                parse("alter table tb1 modify primary key (c, a) not enforced", SqlDialect.DEFAULT);
+        assertAlterTableSchema(
+                operation,
+                ObjectIdentifier.of("cat1", "db1", "tb1"),
+                Schema.newBuilder()
+                        .column("a", DataTypes.STRING().notNull())
+                        .column("b", DataTypes.BIGINT().notNull())
+                        .column("c", DataTypes.BIGINT().notNull())
+                        .primaryKey("c", "a")
+                        .build());
+    }
+
+    @Test
+    public void testAlterTableSchemaModifyWatermark() throws Exception {
+        prepareNonManagedTable(false);
+
+        // test modify watermark
+        Operation operation =
+                parse(
+                        "alter table tb1 modify watermark for a as to_timestamp(a) - interval '1' minute",
```

Comment:
The base doesn't define watermark, so this is equivalent to adding a watermark, I think we should check whether the table has defined a watermark before modifying it?